### PR TITLE
Unified pkg/bootconfig/BootConfig and localboot/BootConfig

### PR DESCRIPTION
**Pull request by insomniacslk**
There was duplication across these two packages, so I merged them into one BootConfig

### CODE REVIEW

File: localboot/grub.go [hunk 1]
```diff
@@ -78,21 +79,11 @@ func ParseGrubCfg(grubcfg string, basedir string, grubVersion int) []BootConfig
...
 					cmdline = strings.Replace(cmdline, `\$`, "$", -1)
 				}
-				fullpath := path.Join(basedir, kernel)
-				fd, err := os.Open(fullpath)
-				if err != nil {
-					debug("error opening kernel file %s: %v", fullpath, err)
-				}
-				cfg.Kernel = fd
-				cfg.Cmdline = cmdline
+                cfg.Kernel = path.Join(basedir, kernel)
```
Comment (zaolin) [inline localboot/grub.go]: indentation

File: localboot/main.go [hunk 2]
```diff
@@ -134,8 +135,7 @@ func BootGrubMode(devices []storage.BlockDev, baseMountpoint string, guid string
 	for _, cfg := range bootconfigs {
 		debug("Trying boot configuration %+v", cfg)
 		if err := cfg.Boot(); err != nil {
-			log.Printf("Failed to boot kernel %s: %v", cfg.Kernel.Name(), err)
-			cfg.Close()
```
Comment (zaolin) [inline localboot/main.go:138]: ?

File: pkg/bootconfig/bootconfig.go [hunk 3]
```diff
@@ -13,13 +20,49 @@ type BootConfig struct {
...
+        if kernel != nil {
+            if err := kernel.Close(); err != nil {
+                log.Printf("Error closing kernel file descriptor: %v", err)
+            }
+        }
+        if initramfs != nil {
+            if err := initramfs.Close(); err != nil {
+                log.Printf("Error closing initramfs file descriptor: %v", err)
+            }
+        }
```
Comment (zaolin) [inline pkg/bootconfig/bootconfig.go]: Don't forget the devicetree

File: localboot/grub.go (hunk 1 above)
Comment (insomniacslk) [reply localboot/grub.go]: darn, I committed from a machine without the goimports plugin -.-

File: localboot/main.go (hunk 2 above)
Comment (insomniacslk) [reply localboot/main.go:138]: this is not necessary anymore, the fields are all strings now, not `os.File`, and the cleanup is done in `BootConfig.Boot`

File: pkg/bootconfig/bootconfig.go (hunk 3 above)
Comment (insomniacslk) [reply pkg/bootconfig/bootconfig.go]: it's not necessasry here - the devicetree field is a string, and it's only used with KexecBin. The cleanup is done there

Comment (zaolin) [reply pkg/bootconfig/bootconfig.go]: ah right, kexec file load does not use the dt. Sorry!

File: localboot/main.go (hunk 2 above)
Comment (zaolin) [reply localboot/main.go:138]: okay

File: localboot/grub.go (hunk 1 above)
Comment (insomniacslk) [reply localboot/grub.go]: also.. why doesn't Stickler complain?

**Review by zaolin — APPROVED** [review]

**Comment by insomniacslk** [comment]
I've erroneously committed the kexecbin change along with this one, which was not tested yet and broke kexec if kexec-tools is not in the image :( will try to fix today, or roll-back the kexec change

**Comment by insomniacslk** [comment]
The fix should be https://github.com/u-root/u-root/pull/996

**Comment by insomniacslk** [comment]
second part of the fix at https://github.com/systemboot/systemboot/pull/67

**Pull request merged**